tests/bluetooth/host/keys/bt_keys_add_type/src/main.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zephyr.h>
+#include <bluetooth/bluetooth.h>
+#include "testing_common_defs.h"
+#include <keys.h>
+#include "host_mocks/assert.h"
+#include "mocks/rpa.h"
+#include "mocks/keys_help_utils.h"
+
+/* This LUT contains different combinations of ID, Address and key type.
+ * Item in this list will be used to fill keys pool.
+ */
+static const struct id_addr_type testing_id_addr_type_lut[CONFIG_BT_MAX_PAIRED] = {
+
+	{ BT_ADDR_ID_1, BT_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
+	{ BT_ADDR_ID_1, BT_ADDR_LE_3, BT_KEYS_IRK },
+
+	{ BT_ADDR_ID_2, BT_ADDR_LE_1, BT_KEYS_LTK },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_3, BT_KEYS_IRK },
+	{ BT_ADDR_ID_2, BT_ADDR_LE_2, BT_KEYS_LOCAL_CSRK },
+
+	{ BT_ADDR_ID_3, BT_ADDR_LE_1, BT_KEYS_REMOTE_CSRK },
+	{ BT_ADDR_ID_3, BT_ADDR_LE_2, BT_KEYS_LTK_P256 },
+
+	{ BT_ADDR_ID_4, BT_ADDR_LE_1, BT_KEYS_ALL }
+};
+
+/* Global iterator to iterate over the ID, Address and type LUT */
+static uint8_t params_it;
+
+/* Pointer to the current set of testing parameters */
+struct id_addr_type const *current_params_vector;
+
+/* This list will hold returned references while filling keys pool */
+static struct bt_keys *returned_keys_refs[CONFIG_BT_MAX_PAIRED];
+
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_type_lut) == CONFIG_BT_MAX_PAIRED);
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_type_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Set key type and verify the value
+ *
+ *  Constraints:
+ *   - Full keys pool list
+ *
+ *  Expected behaviour:
+ *   - The key type value is set correctly
+ */
+void test_set_type_value_correctly(void)
+{
+	struct bt_keys *key_ref;
+	int type = current_params_vector->type;
+
+	/* 'params_it' is incremented before running the test case */
+	key_ref = returned_keys_refs[params_it - 1];
+
+	bt_keys_add_type(key_ref, type);
+
+	zassert_true(key_ref->keys == type, "bt_keys_add_type() set incorrect key type value");
+}
+
+/* Initialize test variables */
+static void reset_global_test_round_params(void)
+{
+	params_it = 0;
+	current_params_vector = NULL;
+}
+
+/* Fill the keys pool with testing (ID, Address) pairs */
+static void keys_fill_pool(void)
+{
+	uint8_t id;
+	bt_addr_le_t *addr;
+	struct id_addr_type const *params_vector;
+
+	zassert_true(check_key_pool_is_empty(),
+		     "List isn't empty, make sure to run this test just after a fresh start");
+
+	for (size_t it = 0; it < ARRAY_SIZE(testing_id_addr_type_lut); it++) {
```

Comment:
I would suggest since this module is about the behavior of bt_keys_add_type, this specific list of pool values isn't really relevant. bt_keys_add_type as its written updates the value of a mask in a bt_keys structure, irrespective of where that structure resides. It's reasonable to ensure that a bt_keys value in the pool changes when bt_keys_add_type is used to future-proof the tests, but you could define a bt_keys on the stack and verify that the mask changes correctly without ay of this setup, teardown or iteration. The coverage impact of the test (today) would be the same, assuming you are not relying on this additional testing to fill in blanks on bt_keys_get_addr testing. If so, I'd move those behaviors to the bt_keys_get_addr test.

Reply:
Good one